File: lib/python/src/wg_portal_client/interfaces.py
from ._http import _HttpClient
from ._models import WgInterface


class InterfacesApi:
    """API methods for WireGuard interfaces."""

    def __init__(self, http: _HttpClient) -> None:
        self._http = http

    def get_all(self) -> list[WgInterface]:
        return self._http.get("/interface/all")  # type: ignore[return-value]

    def get_by_id(self, id: str) -> WgInterface:
        return self._http.get(f"/interface/by-id/{id}")  # type: ignore[return-value]

    def prepare(self) -> WgInterface:
        return self._http.get("/interface/prepare")  # type: ignore[return-value]

    def create(self, data: WgInterface) -> WgInterface:
        return self._http.post("/interface/new", data)  # type: ignore[return-value]

    def update(self, id: str, data: WgInterface) -> WgInterface:
        return self._http.put(f"/interface/by-id/{id}", data)  # type: ignore[return-value]

    def delete(self, id: str) -> None:
        self._http.delete(f"/interface/by-id/{id}")

File: lib/rust/src/interfaces.rs
use crate::http::{HttpClient, WgPortalError};
use crate::models::WgInterface;

pub struct InterfacesApi<'a> {
    http: &'a HttpClient,
}

impl<'a> InterfacesApi<'a> {
    pub(crate) fn new(http: &'a HttpClient) -> Self {
        Self { http }
    }

    pub fn get_all(&self) -> Result<Vec<WgInterface>, WgPortalError> {
        self.http.get("/interface/all", None)
    }

    pub fn get_by_id(&self, id: &str) -> Result<WgInterface, WgPortalError> {
        self.http.get(&format!("/interface/by-id/{}", id), None)
    }

    pub fn prepare(&self) -> Result<WgInterface, WgPortalError> {
        self.http.get("/interface/prepare", None)
    }

    pub fn create(&self, data: &WgInterface) -> Result<WgInterface, WgPortalError> {
        self.http.post("/interface/new", data)
    }

    pub fn update(&self, id: &str, data: &WgInterface) -> Result<WgInterface, WgPortalError> {
        self.http.put(&format!("/interface/by-id/{}", id), data)
    }

    pub fn delete(&self, id: &str) -> Result<(), WgPortalError> {
        self.http.delete(&format!("/interface/by-id/{}", id))
    }
}
